### PR TITLE
Changes required to compile with new Asio executors model

### PR DESCRIPTION
**Pull request by madmongo1**
#refs 1994

### CODE REVIEW

**Comment by chriskohlhoff** [comment]
Remaining MSVC issues may be because I do not currently enable `BOOST_ASIO_HAS_WORKING_EXPRESSION_SFINAE` for any version of MSVC. (I just tried enabling it for a Visual Studio 2019 test runner, but it still triggers errors. They are different errors to earlier versions of MSVC though, so there may be workarounds.)

For maximum portability you may need to define the relevant traits, as described in https://github.com/boostorg/beast/pull/1994#issuecomment-651123446.

**Comment by madmongo1** [comment]
> Remaining MSVC issues may be because I do not currently enable `BOOST_ASIO_HAS_WORKING_EXPRESSION_SFINAE` for any version of MSVC. (I just tried enabling it for a Visual Studio 2019 test runner, but it still triggers errors. They are different errors to earlier versions of MSVC though, so there may be workarounds.)
> 
> For maximum portability you may need to define the relevant traits, as described in [#1994 (comment)](https://github.com/boostorg/beast/pull/1994#issuecomment-651123446).

@chriskohlhoff Just to be clear, the errors seen in MSVC are when compiling vanilla asio types, not a custom executor type.

**Comment by madmongo1** [comment]
Example:

```
.\b2 variant=release cxxstd=14 toolset=msvc define=BOOST_ASIO_HAS_WORKING_EXPRESSION_SFINAE  -q libs/beast/test
```

output:
```
...
    call "bin.v2\standalone\msvc\msvc-14.2\msvc-setup.bat"  >nul
 cl /Zm800 -nologo @"bin.v2\libs\beast\test\beast\core.test\msvc-14.2\release\cxxstd-14-iso\threadapi-win32\threading-multi\core.obj.rsp"

...failed compile-c-c++ bin.v2\libs\beast\test\beast\core.test\msvc-14.2\release\cxxstd-14-iso\threadapi-win32\threading-multi\core.obj...
compile-c-c++ bin.v2\libs\beast\test\beast\http.test\msvc-14.2\release\cxxstd-14-iso\threadapi-win32\threading-multi\http.obj
http.cpp
.\boost/asio/execution/blocking.hpp(364): error C2027: use of undefined type 'boost::asio::execution::detail::blocking::possibly_t<0>'
.\boost/asio/execution/blocking.hpp(209): note: see declaration of 'boost::asio::execution::detail::blocking::possibly_t<0>'
.\boost/asio/execution/blocking.hpp(317): note: see reference to class template instantiation 'boost::asio::execution::detail::blocking_t<0>' being compiled
.\boost/asio/execution/blocking.hpp(374): note: see reference to class template instantiation 'boost::asio::execution::detail::blocking_t<I>' being compiled
.\boost/asio/execution/blocking.hpp(365): error C2027: use of undefined type 'boost::asio::execution::detail::blocking::always_t<0>'
.\boost/asio/execution/blocking.hpp(210): note: see declaration of 'boost::asio::execution::detail::blocking::always_t<0>'
.\boost/asio/execution/blocking.hpp(366): error C2027: use of undefined type 'boost::asio::execution::detail::blocking::never_t<0>'
.\boost/asio/execution/blocking.hpp(211): note: see declaration of 'boost::asio::execution::detail::blocking::never_t<0>'
.\boost/asio/execution/blocking_adaptation.hpp(298): error C2027: use of undefined type 'boost::asio::execution::detail::blocking_adaptation::disallowed_t<0>'
.\boost/asio/execution/blocking_adaptation.hpp(171): note: see declaration of 'boost::asio::execution::detail::blocking_adaptation::disallowed_t<0>'
.\boost/asio/execution/blocking_adaptation.hpp(268): note: see reference to class template instantiation 'boost::asio::execution::detail::blocking_adaptation_t<0>' being compiled
.\boost/asio/execution/blocking_adaptation.hpp(307): note: see reference to class template instantiation 'boost::asio::execution::detail::blocking_adaptation_t<I>' being compiled
.\boost/asio/execution/blocking_adaptation.hpp(299): error C2027: use of undefined type 'boost::asio::execution::detail::blocking_adaptation::allowed_t<0>'
.\boost/asio/execution/blocking_adaptation.hpp(172): note: see declaration of 'boost::asio::execution::detail::blocking_adaptation::allowed_t<0>'
.\boost/asio/execution/bulk_guarantee.hpp(369): error C2027: use of undefined type 'boost::asio::execution::detail::bulk_guarantee::unsequenced_t<0>'
.\boost/asio/execution/bulk_guarantee.hpp(210): note: see declaration of 'boost::asio::execution::detail::bulk_guarantee::unsequenced_t<0>'
.\boost/asio/execution/bulk_guarantee.hpp(320): note: see reference to class template instantiation 'boost::asio::execution::detail::bulk_guarantee_t<0>' being compiled
.\boost/asio/execution/bulk_guarantee.hpp(379): note: see reference to class template instantiation 'boost::asio::execution::detail::bulk_guarantee_t<I>' being compiled
.\boost/asio/execution/bulk_guarantee.hpp(370): error C2027: use of undefined type 'boost::asio::execution::detail::bulk_guarantee::sequenced_t<0>'
.\boost/asio/execution/bulk_guarantee.hpp(211): note: see declaration of 'boost::asio::execution::detail::bulk_guarantee::sequenced_t<0>'
.\boost/asio/execution/bulk_guarantee.hpp(371): error C2027: use of undefined type 'boost::asio::execution::detail::bulk_guarantee::parallel_t<0>'
.\boost/asio/execution/bulk_guarantee.hpp(212): note: see declaration of 'boost::asio::execution::detail::bulk_guarantee

-- More  --
...etc...
```

This source file does not define any user-defined executors.

**Comment by madmongo1** [comment]
@chriskohlhoff here are other command lines that produce failures:

```
./b2  variant=release address-model=32 toolset=msvc-14.2  libs/beast/example
./b2  cxxstd=11 variant=release address-model=32 toolset=msvc-14.2  libs/beast/example
./b2  cxxstd=14 variant=release address-model=32 toolset=msvc-14.2  libs/beast/example
./b2  cxxstd=17 variant=release address-model=32 toolset=msvc-14.2  libs/beast/example
./b2  cxxstd=latest variant=release address-model=32 toolset=msvc-14.2  libs/beast/example
```

**Comment by chriskohlhoff** [comment]
Can you please post the contents of this file:

> bin.v2\libs\beast\test\beast\core.test\msvc-14.2\release\cxxstd-14-iso\threadapi-win32\threading-multi\core.obj.rsp

as it contains the actual command line options for the compiler.

**Comment by madmongo1** [comment]
> Can you please post the contents of this file:
> bin.v2\libs\beast\test\beast\core.test\msvc-14.2\release\cxxstd-14-iso\threadapi-win32\threading-multi\core.obj.rsp
> as it contains the actual command line options for the compiler.

```
"libs\beast\test\lib_test.cpp" -Fo"bin.v2\libs\beast\test\msvc-14.2\release\cxxstd-14-iso\link-static\threading-multi\lib_test.obj"    -TP /O2 /Ob2 /W3 /GR /MD /Zc:forScope /Zc:wchar_t /Zc:inline /Gw /Zc:throwingNew /wd4675 /EHs /std:c++14 -c
-DBOOST_ALL_NO_LIB=1
-DBOOST_BEAST_ALLOW_DEPRECATED
-DBOOST_BEAST_SEPARATE_COMPILATION
-DNDEBUG
"-I."
```

**Comment by chriskohlhoff** [comment]
For some reason that response file above is incomplete. Mine looks like:

```
"example\http\client\async\http_client_async.cpp" -Fo"..\..\bin.v2\libs\beast\example\http\client\async\msvc-14.2\release\cxxstd-14-iso\threading-multi\http_client_async.obj"    -TP /O2 /Ob2 /W3 /GR /MD /Zc:forScope /Zc:wchar_t /Zc:inline /Gw /Zc:throwingNew /wd4675 /EHs /std:c++14 /Ob2 /Oi /Ot /bigobj /permissive- -c 
-DBOOST_ALL_NO_LIB=1 
-DBOOST_ASIO_DISABLE_BOOST_ARRAY=1 
-DBOOST_ASIO_DISABLE_BOOST_BIND=1 
-DBOOST_ASIO_DISABLE_BOOST_DATE_TIME=1 
-DBOOST_ASIO_DISABLE_BOOST_REGEX=1 
-DBOOST_ASIO_NO_DEPRECATED=1 
-DBOOST_ASIO_SEPARATE_COMPILATION 
-DBOOST_BEAST_ALLOW_DEPRECATED 
-DBOOST_BEAST_SEPARATE_COMPILATION 
-DBOOST_COROUTINES_NO_DEPRECATION_WARNING=1 
-DNDEBUG 
-D_CRT_SECURE_NO_WARNINGS=1 
-D_SCL_SECURE_NO_WARNINGS=1 
-D_SILENCE_CXX17_ADAPTOR_TYPEDEFS_DEPRECATION_WARNING 
-D_SILENCE_CXX17_ALLOCATOR_VOID_DEPRECATION_WARNING 
-D_WIN32_WINNT=0x0601 
"-I." 
"-I..\.."
```

The problem only occurs with the `/permissive-` option set. Unfortunately the errors don't really explain why this mode doesn't like the code, but perhaps you can disable this option for now.

**Comment by madmongo1** [comment]
> EDIT: I can enable a workaround in `asio/detail/config.hpp` to fix it. This workaround is already used for gcc <= 7.

Oh. Thank goodness. From the documentation at https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=vs-2019:

> By default, the /permissive- option is set in new projects created by Visual Studio 2017 version 15.5 and later versions. It's not set by default in earlier versions. When the option is set, the compiler generates diagnostic errors or warnings when non-standard language constructs are detected in your code. These constructs include some common bugs in pre-C++11 code.

Given the vast number of people who use the Visual studio IDEs I don't think requiring the disabling of this option is possible, let alone desirable.

I'll test the latest commits and get back to you.

**Comment by madmongo1** [comment]
@chriskohlhoff Good news. It looks like it's building.

FYI: Loads of warnings about very long template names (presumably unavoidable with all the new property code?)

And you may want to look check this:

```
.\boost/asio/detail/impl/scheduler.ipp(365): warning C4267: '+=': conversion from 'size_t' to 'long', possible loss of data
compile-c-c++ bin.v2\libs\chrono\build\msvc-14.0\release\address-model-64\cxxstd-14-iso\threading-multi\thread_clock.obj
thread_clock.cpp
```

**Comment by madmongo1** [comment]
@chriskohlhoff ok all looking good except for MSVC being unable to find specialisations for our custom executor. I have your instructions on how to fix that. Will report back when done.

**Comment by madmongo1** [comment]
@chriskohlhoff error:

```
C:\Users\hodge\github\boostorg\boost\boost/asio/execution/detail/as_invocable.hpp(80): error C2143: syntax error: missing ';' before '}'
```

Looks like the second #ifdef is one line too low:

```
#if !defined(BOOST_ASIO_NO_EXCEPTIONS)
    try
    {
#endif // !defined(BOOST_ASIO_NO_EXCEPTIONS)
      execution::set_value(BOOST_ASIO_MOVE_CAST(Receiver)(*receiver_));
      receiver_ = 0;
    }     //// <<<<==== spurious closing brace here
#if !defined(BOOST_ASIO_NO_EXCEPTIONS)
...
```

**Comment by madmongo1** [comment]
https://github.com/boostorg/asio/pull/360

**Comment by madmongo1** [comment]
@chriskohlhoff still getting errors in msvc-14.1 and above

```
compile-c-c++ bin.v2\libs\beast\msvc-14.1\release\address-model-64\cxxstd-17-iso\link-static\threading-multi\test\lib_asio.obj
lib_asio.cpp
.\boost/asio/execution/blocking.hpp(377): error C2027: use of undefined type 'boost::asio::execution::detail::blocking::possibly_t<0>'
.\boost/asio/execution/blocking.hpp(222): note: see declaration of 'boost::asio::execution::detail::blocking::possibly_t<0>'
.\boost/asio/execution/blocking.hpp(330): note: see reference to class template instantiation 'boost::asio::execution::detail::blocking_t<0>' being compiled
.\boost/asio/execution/blocking.hpp(387): note: see reference to class template instantiation 'boost::asio::execution::detail::blocking_t<I>' being compiled
.\boost/asio/execution/blocking.hpp(378): error C2027: use of undefined type 'boost::asio::execution::detail::blocking::always_t<0>'
.\boost/asio/execution/blocking.hpp(223): note: see declaration of 'boost::asio::execution::detail::blocking::always_t<0>'
```

**Comment by chriskohlhoff** [comment]
N.B. I haven't merged these fixes into the boost source tree yet.

**Comment by chriskohlhoff** [comment]
Now on develop.

**Comment by madmongo1** [comment]
@chriskohlhoff Tests passed for the "allow net-ts executors" case. Will get this PR tidied up and then work on the NO_TS_EXECUTORS version.
Many thanks for your herculean efforts.

**Comment by madmongo1** [comment]
@chriskohlhoff bug in co_spawn:

```
In file included from /home/rhodges/github/boostorg/boost/libs/beast/example/http/server/fast/http_server_fast.cpp:21:
In file included from /home/rhodges/github/boostorg/boost/boost/asio.hpp:47:
In file included from /home/rhodges/github/boostorg/boost/boost/asio/co_spawn.hpp:469:
/home/rhodges/github/boostorg/boost/boost/asio/impl/co_spawn.hpp:60:16: error: no template named 'executor_work_guard'; did you mean 'beast::detail::execution_work_guard'?
    >::type> : executor_work_guard<Executor>
               ^
/home/rhodges/github/boostorg/boost/libs/beast/include/boost/beast/core/detail/work_guard.hpp:34:8: note: 'beast::detail::execution_work_guard' declared here
struct execution_work_guard
       ^
```

`clang-9++ -std=c++20 -fcoroutines-ts -DBOOST_ASIO_NO_TS_EXECUTORS`

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+* Deprecate asio invocation hooks
```
Comment (vinniefalco) [inline CHANGELOG.md]: Handler invoke and allocation hooks are deprecated

File: test/doc/core_1_refresher.cpp [hunk 2]
```diff
@@ -288,7 +288,7 @@ struct associated_allocator<handler, Allocator>
 template<class Executor>
 struct associated_executor<handler, Executor>
 {
-    using type = boost::asio::executor;
+    using type = boost::asio::any_io_executor;
```
Comment (vinniefalco) [inline test/doc/core_1_refresher.cpp:291]: This should be `net::any_io_executor`

File: test/beast/websocket/ssl.cpp [hunk 3]
```diff
@@ -62,7 +62,7 @@ class ssl_test : public unit_test::suite
             boost::asio::ssl::stream<
                 boost::asio::basic_stream_socket<
                     boost::asio::ip::tcp,
-                    boost::asio::executor>>>();
+                    boost::asio::any_io_executor>>>();
```
Comment (vinniefalco) [inline test/beast/websocket/ssl.cpp:65]: `net::any_io_executor`

File: test/beast/core/test_handler.hpp [hunk 4]
```diff
@@ -150,6 +151,7 @@ struct legacy_handler
             asio_handler_is_continuation(&h);
             BEAST_EXPECT(hook_invoked);
         }
+#endif // !defined(BOOST_ASIO_NO_DEPRECATED)
```
Comment (vinniefalco) [inline test/beast/core/test_handler.hpp:170]: I don't put the trailing comment on ifdefs. It is too much hassle to repeat the expression, which is sometimes long and cumbersome. But that's just my preference.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,4 @@
+* Update bind_handler test to use standard executor form.
```
Comment (vinniefalco) [inline CHANGELOG.md:2]: This is not really a user-visible change, and it is not necessary to mention it in the change log as it is already in the commit log.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,4 @@
+* Don't test deprecated hooks.
```
Comment (vinniefalco) [inline CHANGELOG.md:3]: This is not really a user-visible change, and it is not necessary to mention it in the change log as it is already in the commit log.

File: include/boost/beast/core/async_base.hpp [hunk 7]
```diff
@@ -186,7 +186,22 @@ class async_base
...
+    /** The type of executor associated with this object.
+
+    If a class derived from @ref async_base is a completion
+    handler, then the associated executor of the derived class will
+    be this type.
+*/
+    using executor_type = typename
+        net::associated_executor<
+            Handler,
+            typename detail::select_work_guard_t<Executor1>::executor_type
```
Comment (vinniefalco) [inline include/boost/beast/core/async_base.hpp:201]: `detail` identifiers should not appear in the documentation, which will happen here

File: include/boost/beast/core/detail/work_guard.hpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    void reset() noexcept
+    {
+        exec_.reset();
+    }
+
+private:
+
+    boost::optional<executor_type> exec_;
```
Comment (vinniefalco) [inline include/boost/beast/core/detail/work_guard.hpp:63]: `ex` and `ex_` are the canonical names for variables that are executors

File: include/boost/beast/core/detail/work_guard.hpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+struct select_work_guard
+<
+    Executor,
+    typename std::enable_if
+    <
+        boost::asio::execution::is_executor<Executor>::value
+#if defined(BOOST_ASIO_NO_TS_EXECUTORS)
+        || boost::asio::is_executor<Executor>::value
+#else
+        && !boost::asio::is_executor<Executor>::value
```
Comment (vinniefalco) [inline include/boost/beast/core/detail/work_guard.hpp:76]: Should these be `net::` instead of `boost::asio::`?

File: test/beast/config.hpp [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+// Copyright (c) 2016-2020 Vinnie Falco (vinnie dot falco at gmail dot com)
+// Copyright (c) 2020 Richard Hodges (hodges.r@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+#ifndef BOOST_BEAST_TEST_CONFIG_HPP
+#define BOOST_BEAST_TEST_CONFIG_HPP
```
Comment (vinniefalco) [inline test/beast/config.hpp:11]: Do we still need this file?

File: test/beast/config.hpp [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+//
+// Configures which tests will run and how
+
+#include <boost/asio/detail/config.hpp>
```
Comment (vinniefalco) [inline test/beast/config.hpp:16]: You aren't allowed to include detail headers from other libraries

File: test/beast/core/async_base.cpp [hunk 12]
```diff
@@ -24,6 +24,7 @@
 #include <boost/asio/write.hpp>
 #include <boost/core/ignore_unused.hpp>
 #include <stdexcept>
+#include <boost/asio/execution/context_as.hpp>
```
Comment (vinniefalco) [inline test/beast/core/async_base.cpp:27]: Asio includes should come right before std includes

File: test/beast/core/async_base.cpp [hunk 13]
```diff
@@ -32,6 +33,57 @@ namespace beast {
...
+    {}
+
+    bool
+    operator==(ex1_type const &) const noexcept
+    { return true; }
+    bool
+    operator!=(ex1_type const &) const noexcept
+    { return false; }
+};
+static_assert(net::execution::can_execute<ex1_type, net::execution::invocable_archetype>::value, "");
```
Comment (vinniefalco) [inline test/beast/core/async_base.cpp:80]: Or you can use `BOOST_STATIC_ASSERT`

File: test/beast/core/test_handler.hpp (hunk 4 above)
Comment (madmongo1) [reply test/beast/core/test_handler.hpp:170]: ok

File: doc/qbk/release_notes.qbk [hunk 14]
```diff
@@ -11,6 +11,56 @@
 
 [/-----------------------------------------------------------------------------]
 
+[heading Boost 1.74]
+
+[*API Changes]
```
Comment (vinniefalco) [inline doc/qbk/release_notes.qbk:16]: There should be a separate heading
```
[*Asio Changes]

The Boost.Asio library upon which Beast depends has significant
changes and improvements which are now used by Beast:

...
```

File: doc/qbk/release_notes.qbk [hunk 15]
```diff
@@ -11,6 +11,56 @@
...
+   `auto w = net::any_io_executor(net::prefer(e, net::outstanding_work.tracked));`
+ - Correspondingly, the equivalent of `w.reset();` is now `w = {};`
+ - Two argument forms of `post`, `defer` and `dispatch` are deprecated. in favour of the
+   unary versions, which require the supplied completion handler to be associated with
+   an executor as-if by `net::bind_executor`.
+ - Allocator-aware forms of Executor::post, defer and dispatch are deprecated. Dispatching a completion handler `f` on
+   an new-style executor `e` with associated allocator `a` is now spelled:
+   `net::execution::execute(net::prefer(e, net::execution::blocking.possibly, net::execution::allocator(a)), std::forward<F>(f));`
+
+ * Asio introduces a new preprocessor macro BOOST_ASIO_NO_TS_EXECUTORS. If this is defined, the deprecated interfaces
```
Comment (vinniefalco) [inline doc/qbk/release_notes.qbk:32]: Backticks for `BOOST_ASIO_NO_TS_EXECUTORS`

File: doc/qbk/release_notes.qbk [hunk 16]
```diff
@@ -11,6 +11,56 @@
...
+   are not available.
+
+ * Boost.Asio 1.73 behaviour can be retained by not defining BOOST_ASIO_NO_DEPRECATED
+
+[*Changes Required]
+
+* Boost.Beast is tested with both BOOST_ASIO_NO_DEPRECATED set and with BOOST_ASIO_NO_TS_EXECUTORS both set and
```
Comment (vinniefalco) [inline doc/qbk/release_notes.qbk:39]: We can just call it Beast here

File: doc/qbk/release_notes.qbk [hunk 17]
```diff
@@ -61,6 +111,11 @@
 * Fix `echo-op` test
 * Fix non-msvc cmake
 
+[*Miscellaneous]
+
+* Additional tests have been added to ensure correct integration with C++20
+  coroutines when avaialable.
```
Comment (vinniefalco) [inline doc/qbk/release_notes.qbk:117]: Why is this in the 1.73 notes?

**Comment by chriskohlhoff** [comment]
The code in `boost/beast/websocket/impl/stream_impl.hpp` only works with the standard executors. A change like the following fixes it:

```cpp
--- a/include/boost/beast/websocket/impl/stream_impl.hpp
+++ b/include/boost/beast/websocket/impl/stream_impl.hpp
@@ -128,8 +128,8 @@ struct stream<NextLayer, deflateSupported>::impl_type
             boost::empty_init_t{},
             std::forward<Args>(args)...)
         , detail::service::impl_type(
-            net::query(this->boost::empty_value<NextLayer>::get().get_executor(),
-                net::execution::context))
+            this->get_context(
+                this->boost::empty_value<NextLayer>::get().get_executor()))
         , timer(this->boost::empty_value<NextLayer>::get().get_executor())
     {
         timeout_opt.handshake_timeout = none();
@@ -477,6 +477,22 @@ struct stream<NextLayer, deflateSupported>::impl_type
     }

 private:
+    template<class Executor>
+    static net::execution_context&
+    get_context(Executor const& ex,
+        typename std::enable_if< net::execution::is_executor<Executor>::value >::type* = 0)
+    {
+        return net::query(ex, net::execution::context);
+    }
+
+    template<class Executor>
+    static net::execution_context&
+    get_context(Executor const& ex,
+        typename std::enable_if< !net::execution::is_executor<Executor>::value >::type* = 0)
+    {
+        return ex.context();
+    }
+
     bool
     is_timer_set() const
     {
```

You might consider adding a build with `BOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT` defined to your test matrix, as it will help catch problems like this. (This define gives users the option of 100% backward compatibility, in case they need it.)

**Comment by chriskohlhoff** [comment]
> bug in co_spawn:

Now fixed on boostorg/asio develop branch.

File: include/boost/beast/websocket/impl/stream_impl.hpp [hunk 18]
```diff
@@ -128,8 +128,7 @@ struct stream<NextLayer, deflateSupported>::impl_type
             boost::empty_init_t{},
             std::forward<Args>(args)...)
         , detail::service::impl_type(
-            net::query(this->boost::empty_value<NextLayer>::get().get_executor(),
-                net::execution::context))
+            get_context(this->boost::empty_value<NextLayer>::get().get_executor()))
```
Comment (chriskohlhoff) [inline include/boost/beast/websocket/impl/stream_impl.hpp:131]: N.B. In my local changes, I qualified this call with `this->` to ensure ADL wouldn't be used.

**Pull request closed without merging**